Step 4 구현, Step 3 피드백 반영

dahun-lee-daji opened the pull request:
- step 4 및 피드백 반영 완료하여 PR 보내드립니다!

반성 : 
- 요구사항에 맞는 pokergame을 정신없이 구현하다보니, method를 일단 만들기만 해서 구현한 느낌입니다...
- 정신없음에 따라서 커밋을 분배하지 못하고 거대해졌습니다..

고민 할 점 : 
- 상속 시 private한 프로퍼티는 접근 할 수 없는데, 이로 인해 접근하기 위한 method를 만들어 주어야 하는데, 이게 옳은 방법인지 고민 중 입니다.
- Deck을 각 player, dealer의 핸드로 재사용했는데, Hand를 새로 만들었어야 할 지 고민했지만, 큰 의미가 없을거라 생각하여 Deck을 사용했습니다. 하지만 이렇게 될 경우 읽는 사람의 혼란을 초래 할 수 있다고 생각합니다.

dahun-lee-daji committed:
피드백 반영 : 덱 내부 카드 출력

Deck의 CustomStringConvertible protocol을 제거.
string에 + 연산자를 지양하는 방식 채용한 Deck.printCards() 메소드 추가

dahun-lee-daji committed:
피드백 반영 : shuffle()구현, 타입 내부 에러 구현

shuffle()을 직접 구현 : Knuth
TroubleShooter 내부의 Deck 관련 error 분리.
-> TroubleShooter의 필요성? : app의 전체 error를 관리.

dahun-lee-daji committed:
Modify Deck, add new .swift File

Deck을 손 패로 사용 할 수 있도록 수정.
Deck.init 수정 -> 빈 덱을 생성
Deck.reset() 삭제
Deck.newDeck() 구현
Player, pokerGame 타입 구현을 위한 파일 추가

dahun-lee-daji committed:
model 폴더 생성 player, players, dealer 구현

player 손 패 : Deck
카드 받기 기능 구현
dealer : player
카드 주기 기능 구현
players 를 통해 dealer 제외 게이머를 통제

dahun-lee-daji committed:
Deck 수정, player, players, dealer, pokergame 구현

test를 위한 pokergame 구현 및 test코드 작성
deck : pokergame reset을 위해 수정
player : print, reset, 카드 반환 기능 추가
dealer : shuffle, reset, print, 시작 패 분배 기능 추가
시작 패 분배를 위해 sendCard() 리팩토링
players : 딜러를 제외한 player를 print, reset하는 기능 추가

dahun-lee-daji committed:
카드가 부족한 경우 안내 메세지 추가, Update README.md

catch로 인하여 test 중 카드가 부족하면 작동을 멈춘다.
이 때 카드 부족 메세지를 보낸다.

dahun-lee-daji commented:
질문사항 : 
이전 feedback에 error를 nested를 사용해서 error가 발생하는 class내부에 넣어주는게 좋다고 하셨습니다. 
그런데 nested와 관련된 키워드를 탐색해 보았으나, 적합한 정보를 찾지 못했습니다. 추가적인 키워드를 주실 수 있을까요?

godrm commented on PokerGameApp/PokerGameApp/model/Dealer.swift:
```diff
@@ -0,0 +1,59 @@
+    }
+    
+    override func printSelf() {
+        print("딜러 [\(showHandAsString())]")
```
우리가 구현하는 로직에 print를 호출하는 부분이 있으면 이상한 겁니다. 리턴 타입을 활용해서 데이터를 전달해주려고 노력해보세요.

godrm commented on PokerGameApp/PokerGameApp/model/Dealer.swift:
```diff
@@ -0,0 +1,59 @@
+        }
+        catch{
+            print(error)
+            print("카드가 부족하여 종료합니다.")
```
예외 상황이 되면 print 하는게 목적이 아니라 흐름을 제어하는 게 목적입니다. 
sendCard()를 호출한 상위 모듈에서는 정상적으로 나눠줬는지, 실패했는지 알수가 없게 됩니다.
리턴하거나 rethrow해서 상황을 알려주고, 재시도할 지, Alert를 표시할 지 상위에서 판단할 수 있도록 구현하는 게 좋습니다.
print는 데이터 흐름에서 마지막 단계를 처리하는 타입에서만 하면 됩니다.

godrm commented on PokerGameApp/PokerGameApp/model/Deck.swift:
```diff
@@ -0,0 +1,90 @@
+        for index in 0..<self.count() {
+            print(cards[index], terminator: " ")
+        }
+        print("")
```
마찬가지로 출력하는 로직을 포함하지 마세요.

godrm commented on PokerGameApp/PokerGameApp/model/Deck.swift:
```diff
@@ -0,0 +1,90 @@
+        for card in cards {
+            strArray.append(card.description)
+        }
+        return strArray
```
모든 카드를 다 전달하는 메소드가 있으면 내부 속성을 감추고 복잡성을 감추는 의미가 퇴색됩니다.
저 값들을 모두 주는 건 메모리 복사가 일어나기 때문에도 좋은게 아닙니다. 
다른 방법으로는 클로저를 매개변수로 전달해서 내부의 값을 가지고 동작하도록 하는 방법도 있습니다. 참고하세요.

godrm commented on PokerGameApp/PokerGameApp/model/Players.swift:
```diff
@@ -0,0 +1,40 @@
+        for (index,player) in players.enumerated() {
+            print("참가자#\(index) ", terminator : "")
+            player.printSelf()
+        }
```
print 하는 로직을 내부에 두지 말고 외부에서 print 하는 closure를 매개변수로 전달하도록 개선해보세요.

godrm commented on PokerGameApp/PokerGameApp/model/PokerGame.swift:
```diff
@@ -0,0 +1,52 @@
+        self.currentGameStyle = gameStyle
+        self.currentPlayerCount = playerCount
+        self.dealer.resetSelf()
+        self.players.resetSelf()
```
game에서 출력을 위한 객체로 데이터를 전달하는 흐름이 좋습니다.
왜냐하면 앱을 만들때는 UIView 객체가 따로 있을꺼라서 거기서 출력을 해야 합니다.
그 구조를 그대로 활용할 수 있도록 하는게 좋습니다.

dahun-lee-daji replied:
해당 내용은 step 5 에 추가하여 진행하겠습니다

godrm reviewed (changes requested):
다음 피드백을 확인해보세요. 
nested는 수업 시간에도 언급했었습니다. https://docs.swift.org/swift-book/LanguageGuide/NestedTypes.html 읽어보세요.
PR 열어놓을테니 print 하는 로직들 흐름만 개선해서 PUSH해주세요.

dahun-lee-daji committed:
print 로직 변경, legacy code 삭제

Deck :
CustomStringConvertible 채택
print 로직 변경
상위 모델에서 Deck의 card를 print 하는데 쓰인 method 삭제

Player :
CustomStringConvertible 채택
init 수정
print 로직 변경

Dealer :
sendCard()  실행 시 bool을 반환하여 상위 모듈에 결과 알리도록 수정
init 수정

Players :
print 로직 변경

PokerGame :
print 기능 추가

dahun-lee-daji commented:
주말에 PR드려 죄송합니다.😢
피드백 주신, print 관련 로직을 수정했습니다.

godrm commented:
수고하셨습니다.
